buildSrc/src/main/java/org/elasticsearch/gradle/internal/rest/compat/RestCompatTestTransformTask.java

```diff
@@ -0,0 +1,151 @@
+        testPatternSet = getPatternSetFactory().create();
+        testPatternSet.include("/*" + "*/*.yml"); // concat these strings to keep build from thinking this is invalid javadoc
+        transformations = Collections.singletonList(new InjectHeaders(headers));
+    }
+
+    @Inject
+    protected Factory<PatternSet> getPatternSetFactory() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Input
+    public String getSourceSetName() {
+        return sourceSetName;
+    }
+
+    @OutputDirectory
+    public File getOutputDir() {
+        return new File(
+            getSourceSet().orElseThrow(() -> new IllegalArgumentException("could not find source set [" + sourceSetName + "]"))
+                .getOutput()
+                .getResourcesDir(),
+            REST_TEST_PREFIX
+        );
+    }
+
+    @SkipWhenEmpty
+    @InputFiles
+    public FileTree getTestFiles() {
+        return getProject().files(
+            new File(
+                new File(
+                    getSourceSet().orElseThrow(() -> new IllegalArgumentException("could not find source set [" + sourceSetName + "]"))
+                        .getOutput()
+                        .getResourcesDir(),
+                    inputResourceParent
+                ),
+                REST_TEST_PREFIX
+            )
+        ).getAsFileTree().matching(testPatternSet);
+    }
+
+    @TaskAction
+    public void transform() throws IOException {
+        RestTestTransformer transformer = new RestTestTransformer();
+        for (File file : getTestFiles().getFiles()) {
+            YAMLParser yamlParser = yaml.createParser(file);
+            List<ObjectNode> tests = mapper.readValues(yamlParser, ObjectNode.class).readAll();
+            List<ObjectNode> transformRestTests = transformer.transformRestTests(new LinkedList<>(tests), transformations);
+            // convert to url to ensure forward slashes
+            String[] testFileParts = file.toURI().toURL().getPath().split(REST_TEST_PREFIX);
+            assert testFileParts.length == 2; // before and after the REST_TEST_PREFIX
+            File output = new File(getOutputDir(), testFileParts[1]);
+            boolean mkdirs = output.getParentFile().mkdirs();
+            assert mkdirs : "could not make directories for " + output;
+            try (SequenceWriter sequenceWriter = mapper.writer().writeValues(output)) {
+                for (ObjectNode transformedTest : transformRestTests) {
+                    sequenceWriter.write(transformedTest);
+                }
+            }
+        }
+    }
+
+    private Optional<SourceSet> getSourceSet() {
+        Project project = getProject();
+        return project.getConvention().findPlugin(JavaPluginConvention.class) == null
+            ? Optional.empty()
+            : Optional.ofNullable(GradleUtils.getJavaSourceSets(project).findByName(getSourceSetName()));
+    }
+
+    public void setSourceSetName(String sourceSetName) {
```

Comment:
Let's just have this take a `SourceSet` itself instead of a `String` and we can avoid the complexity in `getSourceSet()` altogether. We can then set this from the plugin (where this logic _should_ live), and given the plugin transitively applies the java plugin we can avoid this empty check.

Reply:
Per the comment below, this task no longer cares about sourcesets. Thanks for the tip...i think this makes the task much simpler to reason about. I will look to apply that kind of logic to other tasks in a follow up PR.